### PR TITLE
Remove HIP workarounds in CMake

### PR DESCRIPTION
**Pull request by masterleinad**
Fixes https://github.com/kokkos/kokkos/issues/2791.
I checked that I can configure and cross-compile just fine without these.

Also, add `HIP` to the list of known compilers.

As a side-note: If you forget to set an architecture when cross-compiling you get nasty run time errors like:
```
[==========] Running 157 tests from 3 test cases.
[----------] Global test environment set-up.
[----------] 121 tests from hip
[ RUN      ] hip.atomic_operations_int
./core/unit_test/KokkosCore_UnitTest_HIP: error: node003: task 0: Segmentation fault
```
We might want to force setting an architecture as we do in `CUDA` even though auto-detection seems to work fine when not cross-compiling.

### CODE REVIEW

File: cmake/kokkos_functions.cmake [hunk 1]
```diff
@@ -764,7 +764,7 @@ FUNCTION(kokkos_link_tpl TARGET)
 ENDFUNCTION()
 
 FUNCTION(COMPILER_SPECIFIC_OPTIONS_HELPER)
-  SET(COMPILERS NVIDIA PGI XL DEFAULT Cray Intel Clang AppleClang GNU)
+  SET(COMPILERS NVIDIA PGI XL DEFAULT Cray Intel Clang AppleClang GNU HIP)
```
Comment (dalg24) [inline cmake/kokkos_functions.cmake:767]: Why would you list HIP here?

Comment (masterleinad) [reply cmake/kokkos_functions.cmake:767]: Otherwise, `HIP` is just ignored when calling `COMPILER_SPECIFIC_FLAGS`, e.g., for setting warning flags.

Comment (dalg24) [reply cmake/kokkos_functions.cmake:767]: I don't see that we register any.  Did I miss something?

Comment (masterleinad) [reply cmake/kokkos_functions.cmake:767]: Maybe, my choice of words was not the best. I was only talking about adding to `HIP` to the list of compilers handled here. We have https://github.com/kokkos/kokkos/blob/82e27afc4d51a6b7917650173adb0163939071a5/cmake/kokkos_functions.cmake#L778-L791
in this function. Thus, we don't set any flag if the compiler id doesn't match anything in the list.

**Review by dhollman — APPROVED** [review]
LGTM

**Pull request merged**